# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from [_Think Python_](http://greenteapress.com/wp/think-python-2e/).  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from Think Python deeply.

For more information on using the Jupyter notebook, check out the documentation [here](https://www.datacamp.com/community/tutorials/tutorial-jupyter-notebook##UseJupyter).

Additionally, in order to write text in this document, you will be making use of Markdown cells. A handy cheatsheet for Markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

Reading: _Think Python_ chapters 1-3. _Think Python_ [chapter 5](http://greenteapress.com/thinkpython2/html/thinkpython2006.html), sections 5.2-5.7 and section 5.11.

**Before you begin**: Please log the amount of time you spend on the readings and on this reading journal. One of this last questions in this notebook asks for this.

## Chapter 1
Note: the exercise numbers below match up with the reading for ease of cross referencing.

### Exercise 1.4

~~Start the Python interpreter and use it as a simple calculator.~~ Use a Jupyter notebook as a simple calculator. Enter into each of the following cells a Python expression that calculates the answer to each of the following questions.

* [Example] How many seconds are there in 42 minutes?

In [1]:
42 * 60

2520

* How many seconds are there in 42 minutes 42 seconds?

In [1]:
42*60+42

2562

* How many miles are there in 10 kilometers? Hint: there are 1.61 kilometers in a mile.

In [2]:
10/1.61

6.211180124223602

* If you run a 10 kilometer race in 42 minutes 42 seconds, what is your average pace (time per mile in minutes and seconds)? What is your average speed in miles per hour?

In [11]:
>>> miles=10/1.61

In [13]:
>>> secs_ran=42*60+42

In [21]:
>>> avg_pace=(secs_ran/miles)/60

In [25]:
print(avg_pace)

6.874700000000001


In [23]:
>>> avg_pace_secs_ran=0.8747*60

In [26]:
print (avg_pace_secs_ran)

52.482


Avergae Pace (time per mile in minutes and seconds)=6  mins and 52.482 seconds. 

In [27]:
>>> avg_speed_mph= (10/1.61)/(2562/60/60)

In [28]:
print (avg_speed_mph)

8.727653570337614


## Chapter 2

### Question 1

_Think Python_ defines *interactive mode* and *script mode*. Is the Python code that runs in a Jupyter notebook, using interactive mode or script mode? Enter your answer in the cell below.

interactive mode

### Exercise 1

Is each of the following Python statements a *syntax error*, a *runtime error*, a *semantic error*, or *correct*?

For each question, use the first (code) cell to run the program. Enter your answer into the second (Markdown) cell.

In [ ]:
42 = n

syntax error

* Set `x` and `y` each to the value `1`:

In [35]:
x = y = 1

correct

* Set `x` and `y` each to the value `2`:

In [ ]:
x = y = 1

semantic error

* Multiply `x` and `y`:

In [ ]:
x y

syntax error

### Exercise 2.2

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

The volume of a sphere with radius r is $4/3 \pi  r^3$. What is the volume of a sphere with radius 5?

You can approximate pi with 3.14 or import it through the [Python math library](https://docs.python.org/3/library/math.html#constants). (`import` is covered in [Chapter 3.2](http://greenteapress.com/thinkpython2/html/thinkpython2004.html#sec28).)

In [38]:
import math

In [39]:
4/3*math.pi*5**3

523.5987755982989

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [40]:
24.95*0.4+3+(24.95*0.4+0.75)*59

646.0500000000001

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [42]:
>>> secs_ran=(8*60+15)+(3*7*60+12)+(8*60+15)

In [44]:
>>> mins_ran=secs_ran/60

In [45]:
print (mins_ran)

37.7


In [59]:
>>> time_min=52+38-60

In [51]:
>>> time_hour=6+1

In [61]:
print (time_hour,":", time_min,"am")

7 : 30 am


## Chapter 3

### Exercise 3.1

Write a function named `right_justify` that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('monty')
                                                                 monty
```
Hint: Use string concatenation and repetition. Also, Python provides a built-in function called `len` that returns the length of a string, so the value of `len('monty')` is 5.

In [66]:
 def right_justify(s):
         print((70-len(s))*' '+s)

In [67]:
right_justify("monty")

                                                                 monty


### Exercise 3.3

Note: This exercise should be done using only the statements and other features we have learned so far.

1\. Write a function that draws a grid like the following:

```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
Note: There are four empty spaces between the vertical bars.

To print more than one value on a line, you can print a comma-separated sequence of values:

```
print('+', '-')
```

By default, `print` advances to the next line, but you can override that behavior and put a space at the end, like this:

```
print('+', end=' ')
print('-')
```

The output of these statements is '+ -'.  The output from the next print statement would begin on the next line.

In [31]:
def do_twice(f):
    f()
    f()

def do_four(f):
    do_twice(f)
    do_twice(f)

def print_beam():
    print('+ - - - -', end=' ')

def print_post():
    print('|        ', end=' ')

def print_beams():
    do_twice(print_beam)
    print('+')

def print_posts():
    do_twice(print_post)
    print('|')

def print_row():
    print_beams()
    do_four(print_posts)

def print_grid():
    do_twice(print_row)
    print_beams()

print_grid()

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


2\. Write a function that draws a similar grid with four rows and four columns.

In [33]:
def one_four_one(f, g, h):    
    f()
    do_four(g)
    h()

def print_plus():
    print('+', end=' ')

def print_dash():
    print('-', end=' ')

def print_bar():
    print('|', end=' ')

def print_space():
    print(' ', end=' ')

def print_end():
    print()

def nothing():
    "do nothing"

def print1beam():
    one_four_one(nothing, print_dash, print_plus)

def print1post():
    one_four_one(nothing, print_space, print_bar)

def print4beams():
    one_four_one(print_plus, print1beam, print_end)

def print4posts():
    one_four_one(print_bar, print1post, print_end)

def print_row():
    one_four_one(nothing, print4posts, print4beams)

def print_grid():
    one_four_one(print4beams, print_row, nothing)

print_grid()


+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
|         |         |         |         | 
+ - - - - + - - - - + - - - - + - - - - + 


## Chapter 5.1 - 5.7

### Exercise 5.2

Fermat’s Last Theorem says that there are no positive integers $a$, $b$, and $c$ such that $$a^n + b^n = c^n$$ for any values of $n$ greater than 2.

1\. Write a function named `check_fermat` that takes four parameters – $a$, $b$, $c$ and $n$ – and that checks to see if Fermat’s theorem holds. If n is greater than 2 and
$$a^n + b^n = c^n$$ the program should print: "Holy smokes, Fermat was wrong!" Otherwise the program should print: "No, that doesn’t work."

In [75]:
def check_fermat(a, b, c, n):
    if n > 2 and (a**n + b**n == c**n):
        print("Holy smokes, Fermat was wrong!")
    else:
        print("No, that doesn’t work.")

2\. Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses `check_fermat` to check whether they violate Fermat’s theorem.

In [ ]:
def check_numbers():
    a = int(input("Enter a number for a: "))
    b = int(input("Enter a number for b: "))
    c = int(input("Enter a number for c: "))
    n = int(input("Enter a number for n: "))
    return check_fermat(a, b, c, n)

check_numbers()

### Exercise 5.3 

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

1\. Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [ ]:
def is_triangle(a,b,c):
    if a+b>=c or a+c>=b or b+c>=a:
        print ("Yes")
    else:
        print ("No")

2\. Write a function that prompts the user to input three stick lengths, converts them to integers, and uses `is_triangle` to check whether sticks with the given lengths can form a triangle.

In [ ]:
def prompt_number():
    a = int(input("Enter a number for a: "))
    b = int(input("Enter a number for b: "))
    c = int(input("Enter a number for c: "))
        return is_triangle(a, b, c)
prompt_number()

# Supplementary Reading
Read [Holding a Program in One's Head](http://paulgraham.com/head.html). From the list of bullet points:

1\. Is there one that you already apply to any activities in your life? Name the bullet point, and describe the area where you apply it.

Avoid distraction: every time when I do my homework, I dedicate a time slot where I would not be distracted in any cases so I can focus on my work.

2\. Is there a point that you haven't applied but would like to?

Use succinct languages

3\. Is there a point that you don't believe in, don't believe can apply in this course, or doesn't make sense to you? If so, name the bullet point and briefly describe why.

No, I believe all points can be extremely helpful in this course.

# Course Policies
What is the maximum amount of time you should spend on a reading journal before asking for help?

90 mins

Can you get credit for a late reading journal?

Yes

How many reading journals can you turn in late without penalty?

One

Is it okay to use external sources? If so, what do you need to do in order to do this?

Yes if we use appropriate attribution, but for the journals, we should try our best effort to use our own codes as practices. Every code example we turn in must have a top-level comment identifying us as the author and listing anyone who made a substantial contribution. 

## Quick poll
About how long did you spend working on this Reading Journal?

6 hours

## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Slack.  However, consider using this channel if it doesn't make sense to post your note to Slack.

N/A